### ISSUE TITLE
'Error: Model is disposed!` due to timeout not being disposed

### ISSUE THREAD

**Issue by christianvogt** [remcohaszing/monaco-marker-data-provider#1]
There is a timeout [here](https://github.com/remcohaszing/monaco-marker-data-provider/blame/c46a71af9429eec0718d00180a0e59e8756b116e/src/monaco-marker-data-provider.ts#L80-L89) which does not get properly disposed when all listeners are disposed.

As a result, it's possible to get encounter the following error:
```
textModel.js:230 Uncaught (in promise) Error: Model is disposed!
    at St._assertNotDisposed (textModel.js:230:13)
    at St.getVersionId (textModel.js:435:9)
    at s (monaco-marker-data-provider.js:25:27)
    at monaco-marker-data-provider.js:39:33
```

**Comment by remcohaszing** [comment]
Nice find! I think I see what’s going wrong. I’ll try to fix it this week. Could you provide a failing test case? That would be really useful.

Also out of curiosity: What are you using this for?

**Comment by christianvogt** [comment]
Unfortunately I do not have an isolated test case. We use a library which renders a code editor with `@monaco-editor/react` under the covers and I was investigating adding yaml language support through `monaco-yaml`. Once everything was configured, we encountered this issue by simply navigating away from the editor which would unmount the react component. The error didn't cause any real issues but it does result in an uncaught error logged to console. This actually broke our cypress tests which, by default, fails whenever uncaught errors are encountered.

**Comment by remcohaszing** [comment]
Everything you described in the follow-up comment makes perfect sense. Do you also see the error as a user or only in Cypress tests? I expect this only happens if you act really fast, like a browser automation tool would, but not a human.

I haven’t been able to reproduce this in a test though. Could you try replacing `onModelAdd` with the following in your `node_modules` to see if it fixes the error?

```js
const onModelAdd = (model) => {
  if (!matchesLanguage(model)) {
    return
  }

  let handle
  const onDidChangeContent = model.onDidChangeContent(() => {
    clearTimeout(handle)
    handle = setTimeout(() => {
      doValidate(model)
    }, 500)
  })

  listeners.set(model, {
    dispose() {
      clearTimeout(handle)
      onDidChangeContent.dispose()
    }
  })

  doValidate(model)
}
```

**Comment by christianvogt** [comment]
Tested the above and it works.

The issue was reproducible in both cypress and in application. You are right that the user must act fast to see the error.

eg. type in the editor which would result in an error and immediately click away to unmount the component. This needs to occur in < 500ms to yield the error.

**Comment by remcohaszing** [comment]
Fixed in [v1.2.3](https://github.com/remcohaszing/monaco-marker-data-provider/releases/tag/v1.2.3) :tada:
